env0/resource_environment_discovery_configuration.go

```diff
@@ -0,0 +1,349 @@
+				Default:     false,
+			},
+			"ssh_key_id": {
+				Type:         schema.TypeString,
+				Description:  "The ssh key id that will be available during deployment",
+				Optional:     true,
+				RequiredWith: []string{"ssh_key_name"},
+			},
+			"ssh_key_name": {
+				Type:         schema.TypeString,
+				Description:  "The ssh key name that will be available during deployment",
+				Optional:     true,
+				RequiredWith: []string{"ssh_key_id"},
+			},
+			"retries_on_deploy": {
+				Type:             schema.TypeInt,
+				Description:      "number of times to retry when deploy fails (between 1 and 3)",
+				Optional:         true,
+				ValidateDiagFunc: ValidateRetries,
+			},
+			"retry_on_deploy_only_when_matches_regex": {
+				Type:         schema.TypeString,
+				Description:  "retry (on deploy) if error matches the specified regex",
+				Optional:     true,
+				RequiredWith: []string{"retries_on_deploy"},
+			},
+			"retries_on_destroy": {
+				Type:             schema.TypeInt,
+				Description:      "number of times to retry when destroy fails (between 1 and 3)",
+				Optional:         true,
+				ValidateDiagFunc: ValidateRetries,
+			},
+			"retry_on_destroy_only_when_matches_regex": {
+				Type:         schema.TypeString,
+				Description:  "retry (on destroy) if error matches the specified regex",
+				Optional:     true,
+				RequiredWith: []string{"retries_on_destroy"},
+			},
+			"github_installation_id": {
+				Type:        schema.TypeInt,
+				Description: "github repository id",
+				Optional:    true,
+			},
+			"bitbucket_client_key": {
+				Type:        schema.TypeString,
+				Description: "bitbucket client",
+				Optional:    true,
+			},
+			"gitlab_project_id": {
+				Type:         schema.TypeInt,
+				Description:  "gitlab project id",
+				Optional:     true,
+				RequiredWith: []string{"token_id"},
+			},
+			"is_azure_devops": {
+				Type:         schema.TypeBool,
+				Optional:     true,
+				Description:  "set to true if azure devops is used",
+				Default:      false,
+				RequiredWith: []string{"token_id"},
+			},
+			"token_id": {
+				Type:        schema.TypeString,
+				Description: "a token id to be used with 'gitlab' or 'azure_devops'",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func discoveryReadSshKeyHelper(putPayload *client.EnvironmentDiscoveryPutPayload, d *schema.ResourceData) {
+	sshKeyId := d.Get("ssh_key_id").(string)
+	if sshKeyId != "" {
+		sshKeyName := d.Get("ssh_key_name").(string)
+		putPayload.SshKeys = append(putPayload.SshKeys, client.TemplateSshKey{
+			Id:   sshKeyId,
+			Name: sshKeyName,
+		})
+	}
+}
+
+func discoveryWriteSshKeyHelper(getPayload *client.EnvironmentDiscoveryPayload, d *schema.ResourceData) {
+	var sshKey client.TemplateSshKey
+
+	if len(getPayload.SshKeys) > 0 {
+		sshKey = getPayload.SshKeys[0]
+	}
+
+	d.Set("ssh_key_id", sshKey.Id)
+	d.Set("ssh_key_name", sshKey.Name)
+}
+
+func discoveryValidatePutPayload(putPayload *client.EnvironmentDiscoveryPutPayload) error {
```

Comment:
all sorts of validations that cannot be done at the schema level (for various reasons).